### PR TITLE
Interface for the Validation of data.

### PR DESCRIPTION
**Pull request by klees**
Proposal for a interface to be used for restrictions that are checked during validation.

I'm not sure if `Restriction` should be replaced by `Constraint`.

This aims at being used in the input handling of the UI framework to enable type hinting and moving validation related problems out of the UI framework. Timon and I had the feeling that this should be usefull for a lot of other purposes as well, so we propose this as a global interface. A global library also makes it possible to introduce an share some common restrictions in the system.

Have a look [here](https://github.com/klees/ILIAS/blob/trunk_validation/src/Validation/README.md) for examples.

I kindly ask all developers to comment this PR to request changes or provide feedback. After two weeks I will propose to merge this request during JourFixe if more than 2/3 of the commenting core-developers, but at least two, support it. Thx!

### CODE REVIEW

File: src/Validation/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	assert(false); // does not happen
+}
+
+// Let them throw an exception:
+$raised = false;
+try {
+	$lt10->check(20);
+	assert(false); // does not happen
+}
+catch (\InvalidArgumentException $e) {
```
Comment (chfsx) [inline src/Validation/README.md]: Why this throws a InvalidArgumentException? These exceptions are usually thrown if an argument is not of the expected type.

http://php.net/manual/de/class.invalidargumentexception.php 

Maybe a UnexpectedValueException is better, or an own exception.

File: src/Validation/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Validations
+
+This service abstract validations of values and provides some basic validations
```
Comment (chfsx) [inline src/Validation/README.md]: typo abstract*s*

File: src/Validation/Factory.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+	 * Get a custom restriction.
+	 *
+	 * If the provided value !$is_ok will either use the $error (if it is a string)
+	 * or provide the value to the $error callback.
+	 *
+	 * @param   callable          $is_ok MUST return boolean
+	 * @param   string|callable   $error
+	 * @return  Restriction
+	 */
+	public function custom($is_ok, $error);
```
Comment (chfsx) [inline src/Validation/Factory.php]: Why not typehinting on callable for $is_ok? And I'm not sure if it's a good idea to accept both types string|callable for a parameter. Something like that leads to switchs or ifs in the implementaion.

Additionally the possibility to use custom() may lead to hundreds of those custom() usages, maybe it's better to create PR for new types of restrictions.

**Review by chfsx — CHANGES_REQUESTED** [review]
Hi @klees ans thanks a lot for this PR. I also thinks it's the right way to go for forms and event better to provide this interface for other code.

see my comments, I thinks only small things or questions to discuss...

File: src/Validation/Factory.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+	/**
+	 * Get a restriction that sequentially checks the supplied restrictions.
+	 *
+	 * The new restriction tells the problem of the first violated restriction.
+	 *
+	 * @param   Restriction[]   $others
+	 * @return  Restriction
+	 */
+	public function seq(array $others);
```
Comment (Amstutz) [inline src/Validation/Factory.php]: Matter of personal taste: I hate those 3 letter Abbreviation. It is not immediately clear to me, what seq is supposed to be: sequence? sequent? sequently? I think naming can (should) be more precise than that. However, I see also the benefit of having, short readable lines. Up to you.

File: src/Validation/Factory.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+	/**
+	 * Get a restriction that checks the supplied restrictions in parallel.
+	 *
+	 * The new restriction tells the problems of all violated restrictions.
+	 *
+	 * @param   Restriction[]   $others
+	 * @return	Restriction
+	 */
+	public function par(array $others);
```
Comment (Amstutz) [inline src/Validation/Factory.php]: Matter of personal taste: I hate those 3 letter Abbreviation. Here the problem is bigger than above. par to me those not immediatly hint for parellel. Could also be parity or similar.

File: src/Validation/Factory.php (hunk 3 above)
Comment (Amstutz) [reply src/Validation/Factory.php]: I do not have a strong opinion versus allowing string and callable or not. Both have benefits. I slightly prefer accepting both for pure convenience on the consumer side.

I think it is necessary to allow custom validation for very specific scenarios. It is a thought one though. The need for a PR for each new validation could lead to not properly validated code though to the fear of the need to discuss a PR in the JF. However, we should always encourage to add reusable validations here. Other than UI stuff, the use of validations are not immediately visible and therefore easier to just leave out if the overhead gets too big.

File: src/Validation/Restriction.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * A restriction encodes some constraint on values.
+ *
+ * Restrictions MUST NOT modify the supplied value.
+ */
+interface Restriction {
+	/**
+	 * Checks the provided value.
+     *
```
Comment (Amstutz) [inline src/Validation/Restriction.php]: Indentation issue.

File: src/Validation/README.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+// Get to know what the problem with some value is:
+assert(is_string($gt0->problemWith(-10)));
+
+// Combine them in a way that the restrictions are checked one after another:
+$between_0_10 = $f->seq([$gt0, $lt10]);
+
+// Or in a way that they are checked independently:
+$also_between_0_10 = $f->par([$gt0, $lt10]);
+```
```
Comment (Amstutz) [inline src/Validation/README.md]: A further example with problem builder would be nice.

**Review by Amstutz — APPROVED** [review]
Thank you for this Request. I like and approve it. Some minor feedbacks, mostly matter of personal preference.

File: src/Validation/Factory.php (hunk 4 above)
Comment (wanze) [reply src/Validation/Factory.php]: I have the same opinion as Timon, I think it is better to be explicit when naming methods than using abbreviations.

Comment (chfsx) [reply src/Validation/Factory.php]: +1

File: src/Validation/Factory.php (hunk 3 above)
Comment (klees) [reply src/Validation/Factory.php]: Accepting `string|callable` won't lead to switches if the implementor just uses `function($_) use ($error) { return $error; }` to normalize the string to a callable.

I see your point with the custom validations, we should at least somehow encourage to add validations in the lib.

**Comment by klees** [comment]
Thanks for your feedback. Any opinions about Restriction vs. Constraint?

**Comment by Amstutz** [comment]
@klees: about Restriction vs. Constraint:

> "Restriction, constraint, restraint, or limitation?
> These are all things that limit what you can do. A restriction is a rule or law that is made by someone in authority. A constraint is something that exists rather than something that is made, although it may exist as a result of someone's decision. A restraint is also something that exists: it can exist outside yourself, as the result of someone else's decision; but it can also exist inside you, as a fear of what other people may think or as your own feeling about what is acceptable:moral/social/cultural restraints. A limitation is more general and can be a rule that someone makes or a fact or condition that exists."

Source: [http://english.stackexchange.com/questions/92696/constraint-vs-restriction](http://english.stackexchange.com/questions/92696/constraint-vs-restriction)

Going from this, I would chose constraint. But I think the two are very similar.

**Comment by klees** [comment]
@Amstutz: Very nice, thx. That supports my feeling that it should be `Constraint`.

**Comment by klees** [comment]
Hi @chfsx, are there any other changes you would suggest?

**Comment by chfsx** [comment]
currently it has conflicts but beside this everything's fine and thank you very much!

**Review by chfsx — APPROVED** [review]

**Comment by Amstutz** [comment]
@klees Are you still working on this? If not I guess this can be merged.

**Comment by klees** [comment]
Definitely not sick, @shecken is working on the implementation: https://github.com/klees/ILIAS/pull/28

**Comment by matthiaskunkel** [comment]
Jour Fixe, May 08, 2017: We give another 14 days to comment and think about this suggestion. Postponed to Jour Fixe at May 22, 2017.

**Comment by matthiaskunkel** [comment]
Jour Fixe, May 22, 2017: No objections were made since the last Jour Fixe. We ask Richard to merge this pull request.

**Pull request merged**